[RFC] eval.c: NULL and empty string bugs

splinterofchaos opened the pull request:
See: #2286

Commit:
system(): Return an empty string if no output.

fixes #2286

Commit:
eval: do not assume a vval.v_string is nonnull.

In the fallowing functions, use get_tv_string() to safely avoid a NULL
pointer dereference.
  * rpcstart
  * rpcrequest
  * rpcnotify
  * jobstart

fixes #2321

oni-link commented on src/nvim/eval.c:
```diff
@@ -12882,6 +12885,7 @@ static void f_rpcstart(typval_T *argvars, typval_T *rettv)
   uint64_t channel_id = channel_from_job(argv);
 
   if (!channel_id) {
+on_error:
```
`channel_id` is not initialized when we jump to this point and use it later as return value.

splinterofchaos replied:
Oops, I put this commit in the wrong branch! Oh well, I thought it was close enough related anyway.

Good catch :+1: . fixed.

oni-link commented on src/nvim/eval.c:
```diff
@@ -12867,13 +12867,17 @@ static void f_rpcstart(typval_T *argvars, typval_T *rettv)
   if (argsl > 0) {
     for (listitem_T *arg = args->lv_first; arg != NULL; arg = arg->li_next) {
-      argv[i++] = xstrdup((char *)arg->li_tv.vval.v_string);
+      argv[i++] = xstrdup((char *) get_tv_string(&arg->li_tv));
```
Same problem in `jobstart()`, where we assume that argument strings never point to `NULL`.

splinterofchaos replied:
:+1: -- I found a few others, too. Assuming the tests pass, I will probably squash them all into one commit.

Commit:
event: Only process if event_init has been run.

Reported by @fourjay, a codepath that causes event_poll() to run before
event_init() will trigger a segfault as the events list will not have
been initialized. Exiting immediately from event_init() causes nvim to
hang, so just exit before running the events.

fixes #2339

splinterofchaos commented:
Added a commit to fix #2339.

Pyrohh commented:
Regarding the commit messages, IIRC you need to put `closes #[issue number]`, as `fixes #[issue number]` doesn't work (might be wrong here...).

splinterofchaos commented:
Github does respect `fixes #_`, and several other aliases. https://help.github.com/articles/closing-issues-via-commit-messages/

However, that page doesn't mention it, but I thought I remembered reading that "fixes" would only be respected if the author had the permissions to close issues.

Pyrohh commented:
> Github does respect fixes #_, and several other aliases

Ah, my mistake. I remember being able to close issues with commits before I had permission to close them in general, but it seems what you did does indeed work. For me at least, the reference to your commits from this PR in other issues have a little marker next to them, as mentioned here:

https://github.com/blog/1506-closing-issues-via-pull-requests

ZyX-I commented:
@splinterofchaos AFAIK it is not the commits’ author who needs to have permissions, but the person who merges them in. (*Not* when you need to close issue in the other repository obviously.) First time heard about closing issues via PRs though.
